alliance/src/a2sdf/doc/a2sdf.1
.pl -.4
.TH A2SDF 1 "December 09, 2002" "ASIM/LIP6" "CAO\-VLSI Reference Manual"
.SH NAME
.TP
a2sdf
 - alliance to sdf format converter 
.so man1/alc_origin.1
.SH SYNOPSIS
.TP
\fBa2sdf\fP 
[\fBoptions\fP] \fIcells_dir\fP \fIoutputname\fP
.br

.SH DESCRIPTION
\fBa2sdf\fP translates the SXLIB library to the SDF format. Please refer to
http://www.eda.org/sdf/ for more information on SDF format.
.br
Only the SXLIB library has got timing informations at the moment, so it's
the only library we are able to convert.
.br
\fBcells directory\fP
.br
The path to the SXLIB cells directory.
.br
\fBoutput name\fP
.br
Just the name of the file containing the timing information in sdf format.
It should be named with a .sdf extension.
.br

.SH OPTIONS
.PP
\fB-v[ersion]\fP displays the tool's version.
.br
\fB-h[elp]\fP displays a small help message.
.br
\fB-V[erbose]\fP sets the verbose mode on.
.br
.so man1/alc_bug_report.1


alliance/CHANGES
# Alliance VLSI CAD System
#
# Home page          : http://www-asim.lip6.fr/alliance/
# E-mail support     : mailto:alliance-support@asim.lip6.fr
# ftp sites          : ftp://ftp.lip6.fr/lip6/softs/alliance/
#                   or ftp://ftp-asim.lip6.fr/pub/alliance/
#
# $Id: CHANGES,v 1.3 2000/01/19 18:35:30 czo Exp $

--------------------------------------------------------------------------------

ALLIANCE release 4.0 (15/10/99)

1/ New Standard Cells Library : sxlib
   sxlib allows multi-layer over cell routing
   and all transitions are now on rising edge (man sxlib)

2/ Improved Tools :
   - asimut is now a temporal logic simulator. 
     You *must* modify your old pattern files to allow delays (man asimut)
   - syf can now handle vbe statements in .fsm
   - dreal : New viewing options ( like -install for private colormap)
   - graal : New viewing options ( like -install for private colormap)
             Input scale changeable, Hierarchy browser, real view (.cif/.gds)
   
3/ New Tools :
   algue : ALliance Graphic User Environement
   ali   : gives information on Alliance Environnement
   b2f   : FSM Abstraction from .vbe
   k2f   : Translate kiss2 format <-> FSM format
   vasy  : Translate .vhd format (Synopsys) <-> .vbe format 
   xsch  : Netlist viewer
   xfsm  : FSM viewer
   xpat  : patterns (with delay) viewer


--------------------------------------------------------------------------------

ALLIANCE release 3.5 (19/06/98)

1/ Lots of improvments and bug fixes

2/ AP : Now supports ten layers of metal

3/ AL : New support for RC net in losig. Parsers al, spice, vti modified.
   Modified Data Structures :
    - lotrs : added field BULK TRNAME
    - locon : added field PNODE
    - losig : added field PRCN, CAPA deleted
    
   Modified functions (the CAPA field has moved) :
        addlosig, addlotrs, dellosig

4/ RCN : New library for RC support : Resistances on losig and capcitances
   beetwen losig.
   
--------------------------------------------------------------------------------

ALLIANCE release 3.2c (20/03/98)

1/ Ap format :

   Alliance ap format now supports 
   - half grid spacing
   - connectors somewhere else than abutment box.

2/ VERILOG driver :

   Support added for Alliance-Cadence Toolbox

3/ MBK_SCALE_X now defaults to 100

--------------------------------------------------------------------------------

ALLIANCE release 3.2b (15/12/97)

1/ Easy install

   A "configure" script is now available to configure
   Alliance on any UNIX system

2/ New driver :

   Support to VERILOG netlist as been added.
   Only driver exits. This means you can save your 
   netlists to VERILOG format 'vlg'

3/ New names :

  - Logic has been splited in 3 parts,
    . bop   : boolean optimizer (logic -o) 
    . scmap : Std cell mapping (logic -s) 
    . c4map : CCCC mapping (logic -c) 

  - Desb is replaced by yagle

  - Alligator is replaced by fpmap (X4000)

  - Netoptim is replaced glop

4/ Cells libraries

   The tree of directories containing the cells 
   libraries has been simplified

--------------------------------------------------------------------------------

ALLIANCE release 3.2 (17/05/97)


1/ GRAPHICAL PATTERN VIEWER

   In order to see the patterns resulting from a simulation,
   the XPAT tools has been developped.

2/ GRAPHICAL FSM VIEWER

   In order to see the state's graph of an FSM the XFSM tools has 
   been developped.

3/ RECTANGLE LAYOUT VIEWER

   Now, DREAL is also a real layout editor.

--------------------------------------------------------------------------------

ALLIANCE release 3.0 (17/05/94)

1/ FPGA SYNTHESIS

   A logic synthesis tools that maps on FPGA is now available.
   It works for X3000 devices

2/ FLOOR-PLAN ROUTING

   Binaries of the CHEOPS router from BULL are available for sparc.
   
3/ TIMING ANALYSIS

   The static timing analisys tools TAS is finally available.
   It can be targetted to several processes though the use of a technological
   file suffixed `elp'.

4/ RECTANGLE LAYOUT VIEWER

   In order to see the layout resulting from a symbolic to real translation,
   the DREAL tools has been developped.
   
--------------------------------------------------------------------------------
ALLIANCE release 2.0 (14/02/94) versus ALLIANCE 1.2

1/ SYMBOLIC LAYOUT EDITOR

  The symbolic layout editor ALC has been replaced by GRAAL.
  GRAAL provides the same functionnalities than ALC, but is much
  more reliable. GRAAL support both CMOS and GaAs symbolic layout.

> man graal

2/ DESIGN RULE CHECKER

  The Design Rule Checker VERSATIL has been replaced by DRUC.
  DRUC provides the same functionnalities than VERSATIL.
  A hierarchical version will be distributed in the next ALLIANCE release.

> man druc

3/ LOGIC SYNTHESIS

  The logic synthesis tool LOGIC has been strongly improved.
  The new tool NETOPTIM is a gate-level net-list optimiser
  that allows to minimize delays in a synthesized gate net-list.
  The Finite-state-machine synthesizer SYF allows to describe and
  synthesize high complexity FSM (more than 100 states)
  It is possible to describe hierachical FSM using stack  (subroutines).

> man logic
> man syf
> man netoptim

4/ DATA-PATH COMPILER

  FPGEN is a data-path compiler using a dedicated macro-cells library.
  DPR is the place and route tool that creates optimized data-path blocks
  from the gate net-list generated by FPGEN.

> man fpgen
> man dpr

5/ PARAMETERIZED MACRO-CELLS

  Six parametrized generators are part of this release:

> man rsa                # fast adder generator
> man bsg                # barrel shifter generator
> man amg                # multiplier generator
> man rfg                # register file generator
> man grog               # high speed ROM generator
> man rage               # static RAM generator

6/ PROCEDURAL LAYOUT DEBUGGER

  The graphic debugger GENVIEW allows to debug custom blocks
  described with the procedural language GENLIB.

alliance/src/a2sdf/src/a2sdf.cpp
#include <dirent.h>
#include <iostream>
#include <popt.h>
#include "mut.h"
#include "A2Sdf.h"
using namespace std;

#define EXTENSION ".vbe"

static void version(void)
{
    cout << "a2sdf(" PACKAGE ") " VERSION << endl;
    cout << "ALLIANCE " << ALLIANCE_VERSION << endl;
}

int
main(int argc, char** argv)
{
    int showUsage = 0;
    int showVersion = 0;
    int verbose = 0;
    const char* cellsDir = NULL;
    const char* sdfFileName  = NULL; 
    poptOption options[] = {
	{ "help", 'h', POPT_ARG_NONE, &showUsage, 0, "print this message", 0},
        { "version", 'v', POPT_ARG_NONE, &showVersion, 0, "print the version information", 0},
        { "Verbose", 'V', POPT_ARG_NONE, &verbose, 0, "verbose mode", 0},
        { 0, 0, 0, 0, 0, 0, 0 }
    };

    poptContext context = poptGetContext("a2sdf", argc, (const char**)argv, options, 0);
    poptSetOtherOptionHelp(context, " -hvV <cellsdir> <sdffilename> ");

    if (int rc = poptGetNextOpt(context) < -1)
    {
	cerr << "a2sdf: bad argument " << poptBadOption(context, POPT_BADOPTION_NOALIAS) << ": " << poptStrerror(rc) << endl;
        cerr << "Try `"  << argv[0] << " --help' for more information"         << endl;
        return 1;
    }

    if (showVersion)
    {
	version();
	return 0;
    }

    if (showUsage)
    {
	version();
	cout << endl;
	poptPrintHelp(context, stdout, 0);
	return 0;
    }
        
    if ((cellsDir = poptGetArg(context)) == 0 || poptPeekArg(context) == 0)
    {
	poptPrintUsage(context, stderr, 0);
	return 1;
    }

    if ((sdfFileName = poptGetArg(context)) == 0 || poptPeekArg(context) != 0) 
    {
	poptPrintUsage(context, stderr, 0);
	return 1;
    }

    poptFreeContext(context);

    alliancebanner ("A2Sdf", VERSION, "Alliance to SDF format translator", "2002", ALLIANCE_VERSION);


    DIR* dir = opendir (cellsDir);
    if (!dir)
    {
	cerr << "No directory " << cellsDir  << endl;
	return 1;
    }
    
    A2Sdf a2sdf(sdfFileName, verbose);

    int extensionlength = strlen(EXTENSION);

    struct dirent* entry = NULL;

    bool cellFound = false;

    while ((entry = readdir (dir)) != NULL)
    {
	char* filename = entry->d_name;
	int filenamelength = strlen (filename);

	if (filenamelength <= extensionlength) continue;
	/* is extension of filename accepted */
	if (strncmp (filename + filenamelength - extensionlength,
		    EXTENSION, extensionlength)) continue;
	cellFound = true;
	string cellName(filename);
	cellName.erase( cellName.find(EXTENSION));
	a2sdf.CreateTimingCell(cellName.c_str());
    }

    if (cellFound)
	a2sdf.Dump();
    else
    {
	cerr << " o Error: No cell to convert in: "
	    << cellsDir << endl;
	return 1;
    }
	    

    return 0;
}
